Only rate limit POST endpoint in Store API

senadir opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Fixes an issue where PUT requests in Store API where getting rate limited, while we only want to rate limit POST requests. This is because PUT requests were being sent as POST for backward compatibility.

Closes WOOPRD-1391

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Enable Rate limiting in WooCommerce -> Settings -> Advanced -> Enable rate limits in Checkout.
2. As a guest user, go to checkout, and start switching between payments methods
3. Open dev tools, network and wait for the call to happen (there is debouncing in place)
4. You shouldn't get rate limited.
5. Try Placing an order, it should pass.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

Fix issue in which regular Checkout requests were getting counted toward rate limit if rate limiting is enabled.
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

senadir committed:
Only rate limit POST endpoint in Store API

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

ralucaStan commented on plugins/woocommerce/src/StoreApi/Authentication.php:
```diff
@@ -265,6 +264,35 @@ protected function is_request_to_store_api() {
 
+	/**
+	 * Check if the request method is POST and validate X-HTTP-Method-Override header.
+	 * This is used to allow PUT requests to be made, as wp.apiFetch sends PUT over POST with the X-HTTP-Method-Override header.
```
```suggestion
	 * Returns true only for POST requests that are NOT overridden to another method
     * via the X-HTTP-Method-Override header (used by wp.apiFetch for PUT/DELETE)
```

ralucaStan reviewed (changes requested):
I tested the PR and it works as expected, the code looks good, I added a small improvement to the function's doc.

I think we should be improving the tests: for the new function you added but also for rate limiting the checkout endpoint as well. There are no relevant tests in `tests/php/StoreApi/RateLimitsTests.php` and we should update that as well with the case your PR is fixing.

senadir committed:
Refactor is_only_post_request method documentation and add unit test for accurate POST request validation

senadir committed:
fix lint

ralucaStan commented on plugins/woocommerce/src/StoreApi/Authentication.php:
```diff
@@ -265,6 +264,35 @@ protected function is_request_to_store_api() {
+	 * Check if the request method is POST and validate X-HTTP-Method-Override header.
+	 * This is used to allow PUT requests to be made, as wp.apiFetch sends PUT over POST with the X-HTTP-Method-Override header.
+	 * 
+	 * @see https://github.com/wordpress/gutenberg/blob/trunk/packages/api-fetch/src/middlewares/http-v1.ts#L21-L43
+	 *
+	 * Returns true only if:
+	 * - REQUEST_METHOD is POST, and
+	 * - X-HTTP-Method-Override header (if present) is also POST.
+	 *
+	 * @return bool
+	 */
+	private function is_only_post_request() {
+		// Check that REQUEST_METHOD is POST.
+		if ( ! isset( $_SERVER['REQUEST_METHOD'] ) || 'POST' !== $_SERVER['REQUEST_METHOD'] ) {
+			return false;
+		}
+
+		// Check X-HTTP-Method-Override header if it exists - it must also be POST.
+		if ( isset( $_SERVER['HTTP_X_HTTP_METHOD_OVERRIDE'] ) ) {
+			$method_override = strtoupper( sanitize_text_field( wp_unslash( $_SERVER['HTTP_X_HTTP_METHOD_OVERRIDE'] ) ) );
+			if ( 'POST' !== $method_override ) {
+				return false;
+			}
+		}
+
+		return true;
+	}
```
The test was added in this file plugins/woocommerce/tests/php/src/Blocks/StoreApi/RateLimitsTests.php

ralucaStan committed:
Improve test file:

- Improve failure message
- Ensure $_SERVER cleanup even on failure with try-finally block
- Expanded test coverage with additional HTTP method scenarios

ralucaStan committed:
Merge branch 'trunk' into fix/apply-rate-limit

ralucaStan reviewed (approved):
I tested again, and it worked as expected.
I addressed the AI comments that suggested some improvements for the test file in 27ee89da2c2db13b808c39764fe137cee70664bb

ralucaStan committed:
Update rate limiting doc

ralucaStan committed:
Merge branch 'fix/apply-rate-limit' of github.com:woocommerce/woocommerce into fix/apply-rate-limit

ralucaStan committed:
Merge branch 'trunk' into fix/apply-rate-limit

opr commented on docs/apis/store-api/rate-limiting.md:
```diff
@@ -29,7 +29,7 @@ A default maximum of 25 requests can be made within a 10-second time frame. Thes
 ## Methods restricted by Rate Limiting
 
-`POST`, `PUT`, `PATCH`, and `DELETE`
+Only `POST` requests are rate limited. Requests using the `X-HTTP-Method-Override` header (such as `PUT`, `PATCH` requests sent via `wp.apiFetch`) are excluded from rate limiting.
```
I don't really agree with this - the behaviour we're fixing is more like a bug, and I don't think it's reasonable to assume anyone built with this expectation in mind, _and_ that their use case would break because of this change.

opr commented on plugins/woocommerce/src/StoreApi/Authentication.php:
```diff
@@ -265,6 +264,31 @@ protected function is_request_to_store_api() {
+	 *
+	 * @return bool
+	 */
+	private function is_only_post_request() {
```
There is a small bug with this method, if it's a `POST` method, but `HTTP_X_HTTP_METHOD_OVERRIDE` is set, but is an empty string, then it would return `false`. I think this should return true.

ralucaStan replied:
@ralucaStan  How probable do you think this edge case is? I think the method should be defined, so anything that's outside of the allowed values should be ignored. So I'm not sure I agree with returning true, but I'm curious if you have some examples to demonstrate that we need to return true.

opr replied:
I don't think it's probable, but it's still a potential bug that makes sense to address.

> I think the method should be defined, so anything that's outside of the allowed values should be ignored.

Agreed, but if it's outside of the defined values then should it be ignored and fall back to the base method? Therefore an invalid header in a POST request would make it fall back to a POST?

> but I'm curious if you have some examples to demonstrate that we need to return true.

Not sure right now, but in the interest of "doing it right" we should consider: does an empty override header indicate intent to use a different method? I don't feel it does, therefore returning true, if the base method is POST, is correct.

ralucaStan replied:
Handled in 8b3d48e8881f30b32b9b94916404f536b06c40b5

opr reviewed (changes requested):
Mostly looks fine but I did find one small bug, I added an inline comment about it, let me know what you think. The tests are great too, but if you do fix the bug, please add one to cover this:


```php
			'POST request with empty override header'  => array(
				'method'   => 'POST',
				'override' => '',
				'expected' => true,
			),
```

ralucaStan commented on docs/apis/store-api/rate-limiting.md:
```diff
@@ -29,7 +29,7 @@ A default maximum of 25 requests can be made within a 10-second time frame. Thes
 ## Methods restricted by Rate Limiting
 
-`POST`, `PUT`, `PATCH`, and `DELETE`
+Only `POST` requests are rate limited. Requests using the `X-HTTP-Method-Override` header (such as `PUT`, `PATCH` requests sent via `wp.apiFetch`) are excluded from rate limiting.
```
The checkout endpoint GET, POST, PUT/PATCH. So the documentation wasn't accurate. The fact that GET AND PUT were rate-limited was an unintended side effect, because we weren't aware of the X-HTTP-Method-Override header.

ralucaStan committed:
Handle empty string method override

opr reviewed (approved):
Thanks @senadir and @ralucaStan for working on this. I tested it and it worked well for me.

ralucaStan commented on plugins/woocommerce/src/StoreApi/Authentication.php:
```diff
@@ -265,6 +264,31 @@ protected function is_request_to_store_api() {
+	 *
+	 * @return bool
+	 */
+	private function is_only_post_request() {
```
We don't have any proof that developers need to customize the logic as this comment mentions. If the need is reported we can consider the alternatives.

ralucaStan commented on plugins/woocommerce/src/StoreApi/Authentication.php:
```diff
@@ -155,8 +155,7 @@ public function opt_in_checkout_endpoint( $result ) {
 			&& preg_match( '#/wc/store(?:/v\d+)?/checkout#', $GLOBALS['wp']->query_vars['rest_route'] )
-			&& isset( $_SERVER['REQUEST_METHOD'] )
-			&& 'POST' === $_SERVER['REQUEST_METHOD']
+			&& $this->is_only_post_request()
```
The `/checkout` doesn't have any DELETE logic. The current PUT actions on the endpoint, like updating payment methods, can't really be abused.

ralucaStan committed:
Merge branch 'trunk' into fix/apply-rate-limit

ralucaStan commented on plugins/woocommerce/src/StoreApi/Authentication.php:
```diff
@@ -265,6 +264,31 @@ protected function is_request_to_store_api() {
+	 *
+	 * @return bool
+	 */
+	private function is_only_post_request() {
```
I think I already answered this, but we have no proof extensions are doing this. In the case the need is reported we can consider the 2 variants.

ralucaStan commented on plugins/woocommerce/src/StoreApi/Authentication.php:
```diff
@@ -155,8 +155,7 @@ public function opt_in_checkout_endpoint( $result ) {
 			&& preg_match( '#/wc/store(?:/v\d+)?/checkout#', $GLOBALS['wp']->query_vars['rest_route'] )
-			&& isset( $_SERVER['REQUEST_METHOD'] )
-			&& 'POST' === $_SERVER['REQUEST_METHOD']
+			&& $this->is_only_post_request()
```
The [rate-limiting](https://developer.woocommerce.com/docs/apis/store-api/rate-limiting/)feature was intended for POST `/checkout` and Place Order flow for Checkout block. 

> developers who may have built custom monitoring, security plugins, or analytics that depend on the previous rate-limiting behavior for all POST requests.

The change will be announced in the beta release, and we can collect feedback. I don't see a need to rate limit the supported PUT methods on the rate limited endpoints, but I'll also let @senadir chime in
